flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala

```diff
@@ -68,12 +70,35 @@ class GroupAggProcessFunction(
     var accumulators = state.value()
 
     if (null == accumulators) {
+      previous = null
       accumulators = new Row(aggregates.length)
       i = 0
       while (i < aggregates.length) {
         accumulators.setField(i, aggregates(i).createAccumulator())
         i += 1
       }
+    } else {
+      // get previous row
+      if (generateRetraction) {
+        if (null == previous) {
+          previous = new Row(groupings.length + aggregates.length)
+          // previous is used to output retract message, so command of previous will always be
+          // Command.Delete
+          previous.command = Command.Delete
+        }
+        i = 0
+        while (i < groupings.length) {
+          previous.setField(i, input.getField(groupings(i)))
+          i += 1
+        }
+        i = 0
+        while (i < aggregates.length) {
+          val index = groupings.length + i
+          val accumulator = accumulators.getField(i).asInstanceOf[Accumulator]
+          previous.setField(index, aggregates(i).getValue(accumulator))
```

Comment:
reduce a local variable:
previous.setField(groupings.length + i, aggregates(i).getValue(accumulator))

also L118 and L127